SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracerUserInterface.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracerUserInterface.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/Connection/SpatialEventTracer.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialClassInfoManager.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracerUserInterface);
+
+FUserSpanId USpatialEventTracerUserInterface::CreateSpanId(UObject* WorldContextObject)
+{
+	const SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return {};
+	}
+
+	return SpatialGDK::SpatialEventTracer::SpanIdToUserSpanId(EventTracer->CreateSpan().GetValue());
+}
+
+FUserSpanId USpatialEventTracerUserInterface::CreateSpanIdWithCauses(UObject* WorldContextObject, const TArray<FUserSpanId>& Causes)
+{
+	const SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return {};
+	}
+
+	TArray<Trace_SpanId> SpanIds;
+	for (const FUserSpanId& UserSpanIdCause : Causes)
+	{
+		if (!UserSpanIdCause.IsValid())
+		{
+			UE_LOG(LogSpatialEventTracerUserInterface, Warning,
+				   TEXT("USpatialEventTracerUserInterface::CreateSpanIdWithCauses - Invalid input cause"));
+			continue;
+		}
+
+		TOptional<Trace_SpanId> CauseSpanId = SpatialGDK::SpatialEventTracer::UserSpanIdToSpanId(UserSpanIdCause);
+		if (CauseSpanId.IsSet())
+		{
+			SpanIds.Add(CauseSpanId.GetValue());
+		}
+	}
+
+	return SpatialGDK::SpatialEventTracer::SpanIdToUserSpanId(EventTracer->CreateSpan(SpanIds.GetData(), SpanIds.Num()).GetValue());
+}
+
+void USpatialEventTracerUserInterface::TraceEvent(UObject* WorldContextObject, const FUserSpanId& UserSpanId,
+												  const FSpatialTraceEvent& SpatialTraceEvent)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr)
+	{
+		return;
+	}
+
+	if (!UserSpanId.IsValid())
+	{
+		return;
+	}
+
+	TOptional<Trace_SpanId> SpanId = SpatialGDK::SpatialEventTracer::UserSpanIdToSpanId(UserSpanId);
+	if (!SpanId.IsSet())
+	{
+		return;
+	}
+
+	EventTracer->TraceEvent(SpatialTraceEvent, SpanId.GetValue());
+}
+
+void USpatialEventTracerUserInterface::TraceRPC(UObject* WorldContextObject, FEventTracerRPCDelegate Delegate,
+												const FUserSpanId& UserSpanId)
+{
+	SpatialGDK::SpatialEventTracer* EventTracer = GetEventTracer(WorldContextObject);
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	if (!UserSpanId.IsValid())
+	{
+		return;
+	}
+
+	TOptional<Trace_SpanId> SpanId = SpatialGDK::SpatialEventTracer::UserSpanIdToSpanId(UserSpanId);
+	if (!SpanId.IsSet())
+	{
+		return;
+	}
+
+	EventTracer->SpanIdStack.Stack(SpanId.GetValue());
+	Delegate.Execute();
```

Comment:
I feel like this is too intrusive for game code, if event tracing is not enabled then the delegate is not called - is there a way to do this without changing control flow?

Reply:
Since this will be tracing already implemented game flows maybe we should avoid needing to change that code significantly to add tracing

Reply:
We should call the delegate even if even event tracing is disabled. Otherwise my only other solution is to expose adding and poping spanIds from the stack which I don't really like.

Reply:
Perhaps executing the delegate in the if clause above will be enough?
```if (EventTracer == nullptr || !EventTracer->IsEnabled())```

Reply:
I imagine we should call the delegate even if we have ET enabled but somehow mess up the SpanId stuff